### PR TITLE
nixos/snapper: change timer wantedBy to timers.target & add snapshotOnBoot

### LINKED ISSUE #136570: snapper-timeline.timer not started after reboot
### Describe the bug
Snapper stops to take snapshots after a system update and reboot.

### Steps To Reproduce

1. Enable snapper in configuration.nix
2. Enable autoUpgrade and autoUpgrade.allowReboot in configuration.nix
3. Wait for system upgrade and reboot.
4. snapperd.service succeds, but snapper-timeline.timer does not start (see below).
5. snapper will start to take snapshots after `nixos-rebuild switch` or `systemctl start snapper-timeline.timer`

### Expected behavior
snapper-timeline.timer should start without intervention after a reboot.

### Additional context

journalctl -u snapper-timeline.timer

```
Sep 03 04:45:00 univac systemd[1]: time-sync.target: Found ordering cycle on snapper-timeline.timer/stop
Sep 03 04:45:00 univac systemd[1]: time-sync.target: Found dependency on basic.target/stop
Sep 03 04:45:00 univac systemd[1]: time-sync.target: Found dependency on ntpd.service/stop
Sep 03 04:45:00 univac systemd[1]: time-sync.target: Found dependency on time-sync.target/stop
Sep 03 04:45:00 univac systemd[1]: time-sync.target: Job snapper-timeline.timer/stop deleted to break ordering cycle starting with time-sync.target/stop
```

systemctl status snapper-timeline.timer
```
● snapper-timeline.timer - Timeline of Snapper Snapshots
     Loaded: loaded (/nix/store/1n6zpiha9d6pzk4zng4faqdw5axjnbb9-unit-snapper-timeline.timer/snapper-timeline.timer; enabled; vendor preset: enabled)
     Active: inactive (dead)
    Trigger: n/a
   Triggers: ● snapper-timeline.service
       Docs: man:snapper(8)
             man:snapper-configs(5)
```

snapper-timeline.timer is not listed in `systemctl list-units --state=failed`

/etc/nixos/configuration.nix

```
 ...
    snapper = {
      configs = {
        "root" = {
          subvolume = "/";
          extraConfig = ''
            TIMELINE_CREATE="yes"
            TIMELINE_CLEANUP="yes"
            TIMELINE_MIN_AGE="1800"
            TIMELINE_LIMIT_HOURLY="12"
            TIMELINE_LIMIT_DAILY="10"
            TIMELINE_LIMIT_WEEKLY="5"
            TIMELINE_LIMIT_MONTHLY="12"
            TIMELINE_LIMIT_YEARLY="10"
          '';
        };
      };
    };
...
```

journalctl -u snapperd.service

```
-- Boot 28d3d1190312405ea9b1cc74d69f1fdb --
Sep 03 04:58:28 univac systemd[1]: Starting DBus interface for snapper...
Sep 03 04:58:28 univac systemd[1]: Started DBus interface for snapper.
Sep 03 04:59:30 univac systemd[1]: snapperd.service: Succeeded.
```

snapper list

```
    # | Type   | Pre # | Date                             | User | Cleanup  | Description    | Userdata
------+--------+-------+----------------------------------+------+----------+----------------+---------
   0  | single |       |                                  | root |          | current        |         
   1  | single |       | Wed 09 Sep 2020 09:06:30 PM CEST | root |          |                |         
   2  | single |       | Wed 09 Sep 2020 09:16:15 PM CEST | root | timeline | timeline       |         
 509  | single |       | Thu 01 Oct 2020 12:00:03 AM CEST | root | timeline | timeline       |         
1122  | pre    |       | Mon 26 Oct 2020 12:38:03 PM CET  | root |          | before cleanup |         
1248  | single |       | Sun 01 Nov 2020 12:00:03 AM CET  | root | timeline | timeline       |         
1873  | single |       | Tue 01 Dec 2020 12:00:02 AM CET  | root | timeline | timeline       |         
2527  | single |       | Fri 01 Jan 2021 12:00:01 AM CET  | root | timeline | timeline       |         
2625  | single |       | Wed 03 Feb 2021 07:00:04 PM CET  | root | timeline | timeline       |         
3004  | single |       | Mon 01 Mar 2021 12:00:01 AM CET  | root | timeline | timeline       |         
3747  | single |       | Thu 01 Apr 2021 12:00:01 AM CEST | root | timeline | timeline       |         
4467  | single |       | Sat 01 May 2021 12:00:01 AM CEST | root | timeline | timeline       |         
5211  | single |       | Tue 01 Jun 2021 12:00:03 AM CEST | root | timeline | timeline       |         
5931  | single |       | Thu 01 Jul 2021 12:00:00 AM CEST | root | timeline | timeline       |         
6675  | single |       | Sun 01 Aug 2021 12:00:00 AM CEST | root | timeline | timeline       |         
6699  | single |       | Mon 02 Aug 2021 12:00:00 AM CEST | root | timeline | timeline       |         
6867  | single |       | Mon 09 Aug 2021 12:00:00 AM CEST | root | timeline | timeline       |         
7035  | single |       | Mon 16 Aug 2021 12:00:00 AM CEST | root | timeline | timeline       |         
7202  | single |       | Mon 23 Aug 2021 12:00:00 AM CEST | root | timeline | timeline       |         
7231  | single |       | Wed 25 Aug 2021 05:00:02 AM CEST | root | timeline | timeline       |         
7250  | single |       | Thu 26 Aug 2021 12:00:01 AM CEST | root | timeline | timeline       |         
7274  | single |       | Fri 27 Aug 2021 12:00:00 AM CEST | root | timeline | timeline       |         
7298  | single |       | Sat 28 Aug 2021 12:00:00 AM CEST | root | timeline | timeline       |         
7322  | single |       | Sun 29 Aug 2021 12:00:00 AM CEST | root | timeline | timeline       |         
7327  | single |       | Mon 30 Aug 2021 05:00:01 AM CEST | root | timeline | timeline       |         
7346  | single |       | Tue 31 Aug 2021 12:00:00 AM CEST | root | timeline | timeline       |         
7370  | single |       | Wed 01 Sep 2021 12:00:01 AM CEST | root | timeline | timeline       |         
7394  | single |       | Thu 02 Sep 2021 12:00:00 AM CEST | root | timeline | timeline       |         
7411  | single |       | Thu 02 Sep 2021 05:00:02 PM CEST | root | timeline | timeline       |         
7412  | single |       | Thu 02 Sep 2021 06:00:01 PM CEST | root | timeline | timeline       |         
7413  | single |       | Thu 02 Sep 2021 07:00:02 PM CEST | root | timeline | timeline       |         
7414  | single |       | Thu 02 Sep 2021 08:00:00 PM CEST | root | timeline | timeline       |         
7415  | single |       | Thu 02 Sep 2021 09:00:00 PM CEST | root | timeline | timeline       |         
7416  | single |       | Thu 02 Sep 2021 10:00:00 PM CEST | root | timeline | timeline       |         
7417  | single |       | Thu 02 Sep 2021 11:00:01 PM CEST | root | timeline | timeline       |         
7418  | single |       | Fri 03 Sep 2021 12:00:01 AM CEST | root | timeline | timeline       |         
7419  | single |       | Fri 03 Sep 2021 01:00:00 AM CEST | root | timeline | timeline       |         
7420  | single |       | Fri 03 Sep 2021 02:00:02 AM CEST | root | timeline | timeline       |         
7421  | single |       | Fri 03 Sep 2021 03:00:02 AM CEST | root | timeline | timeline       |         
7422  | single |       | Fri 03 Sep 2021 04:00:01 AM CEST | root | timeline | timeline       |    
```

### Notify maintainers
@markuskowa

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
these paths will be fetched (0.05 MiB download, 0.28 MiB unpacked):
  /nix/store/p5lnl4zr45n7mf9kz9w8yz3rqh001b5c-bash-interactive-4.4-p23-dev
copying path '/nix/store/p5lnl4zr45n7mf9kz9w8yz3rqh001b5c-bash-interactive-4.4-p23-dev' from 'https://cache.nixos.org'...
 - system: `"x86_64-linux"`
 - host os: `Linux 5.10.61, NixOS, 21.05.2861.3e09410da09 (Okapi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.15`
 - channels(tmp): `"nixpkgs-20.03.2882.51d115ac89d"`
 - channels(root): `"nixos-21.05.2861.3e09410da09, home-manager-20.09"`
 - channels(ch): `""`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
```

### PR DESCRIPTION
**Pull request by Artturin**
###### Motivation for this change
Closes #136570

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Artturin** [comment]
@helma to test: https://nixos.wiki/wiki/Nixpkgs/Reviewing_changes#Pre-flakes

File: nixos/modules/services/misc/snapper.nix [hunk 1]
```diff
@@ -142,7 +150,7 @@ in
     systemd.timers.snapper-timeline = {
```
Comment (markuskowa) [inline nixos/modules/services/misc/snapper.nix:150]: I think we can remove `systemd.timers.snapper-timeline` and instead add `startAt` to `systemd.services.snapper-timeline`.

File: nixos/modules/services/misc/snapper.nix [hunk 2]
```diff
@@ -155,10 +163,25 @@ in
...
+
+    systemd.services.snapper-boot = lib.optionalAttrs cfg.snapshotOnBoot {
+      description = "Take snapper snapshot of root on boot";
+      inherit documentation;
+      serviceConfig.ExecStart = "${pkgs.snapper}/bin/snapper --config root create --cleanup-algorithm number --description boot";
+      serviceConfig.type = "oneshot";
+      unitConfig.ConditionPathExists = "/etc/snapper/configs/root";
+    };
+
+    systemd.timers.snapper-boot = lib.optionalAttrs cfg.snapshotOnBoot {
```
Comment (markuskowa) [inline nixos/modules/services/misc/snapper.nix:179]: All services should depend `local-fs.target` to ensure the file system is actually mounted?

**Comment by markuskowa** [comment]
CC @ck3d as original author of the module.

File: nixos/modules/services/misc/snapper.nix [hunk 3]
```diff
@@ -155,10 +163,25 @@ in
     systemd.timers.snapper-cleanup = {
       description = "Cleanup of Snapper Snapshots";
       inherit documentation;
-      wantedBy = [ "basic.target" ];
+      wantedBy = [ "timers.target" ];
       timerConfig.OnBootSec = "10m";
       timerConfig.OnUnitActiveSec = cfg.cleanupInterval;
     };
+
+    systemd.services.snapper-boot = lib.optionalAttrs cfg.snapshotOnBoot {
```
Comment (ck3d) [inline nixos/modules/services/misc/snapper.nix:171]: If this service is watend by multi-user.targed then no timer is needed

**Comment by ck3d** [comment]
Since I do not use snapper, I can not approve this change.

File: nixos/modules/services/misc/snapper.nix [hunk 4]
```diff
@@ -136,14 +144,9 @@ in
     systemd.services.snapper-timeline = {
       description = "Timeline of Snapper Snapshots";
       inherit documentation;
+      requires = [ "fs.target" ];
```
Comment (markuskowa) [inline nixos/modules/services/misc/snapper.nix:147]:
```suggestion
      requires = [ "local-fs.target" ];
```

File: nixos/modules/services/misc/snapper.nix [hunk 5]
```diff
@@ -155,10 +158,20 @@ in
     systemd.timers.snapper-cleanup = {
       description = "Cleanup of Snapper Snapshots";
       inherit documentation;
-      wantedBy = [ "basic.target" ];
+      wantedBy = [ "timers.target" ];
+      requires = [ "fs.target" ];
```
Comment (markuskowa) [inline nixos/modules/services/misc/snapper.nix:162]:
```suggestion
      requires = [ "local-fs.target" ];
```

File: nixos/modules/services/misc/snapper.nix [hunk 6]
```diff
@@ -9,6 +9,14 @@ in
 {
   options.services.snapper = {
 
+    snapshotOnBoot = mkOption {
```
Comment (markuskowa) [inline nixos/modules/services/misc/snapper.nix:12]: Why not snapshot all existing configs at boot? Doing only a root snapshot on boot seems to be very specific. The name of the config also does not reflect properly that this is only applies to root.

Comment (Artturin) [reply nixos/modules/services/misc/snapper.nix]: changed the option to snapshotRootOnBoot, if someone wants to implement snapshotting all profiles on boot then they may do it in a separate pr

**Comment by markuskowa** [comment]
@GrahamcOfBorg test snapper

**Pull request merged**